### PR TITLE
Standalone resolver

### PR DESCRIPTION
**Pull request by erikbern**
This PR makes it possible to use the resolver without the app, which
1. Reduces some complexity of the `App` class
2. Should make it possible to clean up `Provider._deploy` so we don't have to use the app at all (makes it possible to deploy a single object without creating an app client-side)

### CODE REVIEW

File: modal/_resolver.py [hunk 1]
```diff
@@ -8,50 +8,80 @@
...
-class Resolver:
+
+class StatusRow:
+    def __init__(self, progress):
+        from ._output import step_progress  # Lazy import to only import `rich` when necessary.
+
+        self._spinner = None
+        self._step_node = None
+        if progress is not None:
+            self._spinner = step_progress()
```
Comment (freider) [inline modal/_resolver.py:21]: Should this be?
```suggestion
            self._spinner = step_progress(progress)
```

File: modal/_resolver.py [hunk 2]
```diff
@@ -8,50 +8,80 @@
     Spinner = TypeVar("Spinner")
     Tree = TypeVar("Tree")
 
+from modal_utils.async_utils import synchronize_apis
 
-class Resolver:
+
+class StatusRow:
+    def __init__(self, progress):
```
Comment (freider) [inline modal/_resolver.py:15]: Perhaps type this as an `Optional[Tree]` - I was confused at first thinking it was the initial message

Comment (erikbern) [reply modal/_resolver.py:15]: 👍

**Review by freider — APPROVED** [review]
lgtm!

File: modal/_resolver.py [hunk 3]
```diff
@@ -8,50 +8,80 @@
...
-            if self._step_node is None:
-                self._spinner = step_progress()
-                self._step_node = self._progress.add(self._spinner)
-            step_progress_update(self._spinner, message)
+        self._client = client
+        self._app_id = app_id
+
+    @property
+    def app_id(self):
+        return self._app_id
```
[the remark this answers is not in the document]
Comment (erikbern) [reply modal/_resolver.py:54]: yeah i was actually planning to do that in the next PR, but i'll do it in this!

File: modal/app.py [hunk 4]
```diff
@@ -71,58 +70,28 @@ def app_id(self) -> str:
...
         unindexed_object_ids = list(
-            set(obj.object_id for obj in self._local_uuid_to_object.values())
+            set(obj.object_id for obj in resolver.objects())
             - set(obj.object_id for obj in self._tag_to_object.values())
         )
         req_set = api_pb2.AppSetObjectsRequest(
             app_id=self._app_id,
             indexed_object_ids=indexed_object_ids,
             unindexed_object_ids=unindexed_object_ids,
-            new_app_state=new_app_state,  # type: ignore
```
Comment (freider) [inline modal/app.py:125]: Ohhh no, I missed this removal in the review :O Was this intentional?

Comment (erikbern) [reply modal/app.py:125]: Pretty sure it was accidental :(

**Pull request merged**